gpii/node_modules/eventLog/src/eventLog.js

```diff
@@ -0,0 +1,396 @@
+        },
+        "{lifecycleManager}.events.onSessionStop": {
+            namespace: "eventLog",
+            func: "{that}.logTimedEvent",
+            args: [
+                "lifecycle",
+                "SessionStop",
+                {userToken: "{arguments}.1.model.userToken"},
+                "{arguments}.1.createTime",
+                "sessionTime"
+            ]
+        }
+    }
+});
+
+/**
+ * The timers for timed events.
+ */
+gpii.eventLog.timers = {};
+
+/**
+ * The outstanding log entries waiting for their duration.
+ */
+gpii.eventLog.pendingLogs = {};
+
+/**
+ * The log file.
+ */
+gpii.eventLog.logFilePath = null;
+
+/**
+ * The installation ID
+ */
+gpii.eventLog.installationID = null;
+
+/**
+ * Gets (and stores) the installation ID.
+ * @param installationId The installation ID.
+ */
+gpii.eventLog.setInstallId = function (installationId) {
+    gpii.eventLog.installationID = installationId;
+};
+
+/**
+ * Returns the actual Date and time
+ */
+gpii.eventLog.getTimestamp = function () {
+    return new Date();
+};
+
+/**
+ * Creates an object for the log. Everything in this object is what will be logged, and the "time" field will be added
+ * later when it is actually logged.
+ *
+ * @param module {String} The part of GPII causing this event.
+ * @param event {String} Name of the event.
+ * @param data [optional] Event specific data.
+ * @param level {Object} [optional] Level of the log, see fluid.logLevelsSpec [FATAL,FAIL,WARN,IMPORTANT,INFO,TRACE].
+ * @return {Object} The log object.
+ */
+gpii.eventLog.createLogObject = function (module, event, data, level) {
+    var eventObject = {
+        module: module || "GPII",
+        event: event,
+        level: level
+    };
+
+    var dataWrapped = gpii.eventLog.wrapData(data);
+    if (dataWrapped !== null) {
+        eventObject.data = fluid.copy(dataWrapped);
+    }
+    return eventObject;
+};
+
+/**
+ * Logs an event.
+ *
+ * If this event is the start of something (like "StartSession"), then fill in the timerId parameter and log the
+ * corresponding "end" event (eg, "StopSession") using logDuration().
+ *
+ * @param module {String} The part of GPII causing this event.
+ * @param event {String} The event name.
+ * @param data {Object} [optional] Event specific data.
+ * @param level {Object} [optional] Level of the log, see fluid.logLevelsSpec [FATAL,FAIL,WARN,IMPORTANT,INFO,TRACE].
+ * @param timerId {String} [optional] Unique identifier for a timer, if this event is the start of a duration.
+ */
+gpii.eventLog.log = function (module, event, data, level, timerId) {
+    var eventObject = gpii.eventLog.createLogObject(module, event, data);
+    if (timerId) {
+        gpii.eventLog.startTimer(timerId);
+    }
+
+    gpii.eventLog.writeLog(level, eventObject);
+};
+
+/**
+ * Logs an event, where the event marks the end of a duration. This will add a "duration" field to the data object with
```

Comment:
I know.

This is a short coming of the current analysis system, where each entry is discrete (see [this comment - GPII-2398](https://issues.gpii.net/browse/GPII-2398?focusedCommentId=27014&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-27014)).

I'd prefer to just blindly write entries to the log as they occur, but if you can't just ask it "get me the next 'thing end' entry after this 'thing start'", then there would need to be a way to associate the start/end entries... a "thing id" - "get me the 'thing end' with the same id of this 'thing start'". 

There's still some effort required to associate the start and ends (and it could even be said it's not the responsibility of the analytics to associate start/ends). It just so happens that with this implementation, the "id" is a tick count and is decorated as "duration".

@amatas - is this fair to say?